org/postgresql/jdbc2/TypeInfoCache.java

```diff
@@ -150,9 +153,14 @@ public synchronized void addCoreType(String pgTypeName, Integer oid, Integer sql
         Character delim = ',';
         _arrayOidToDelimiter.put(oid, delim);
 
-        String pgArrayTypeName = "_" + pgTypeName;
```

Comment:
Not sure I'm understanding you.
I changed code so that _ prefix does not imply array and that suffix [] implies array. Both of those are correct. How is that not better?

Didn't I just show you an example when you have to use [] ?
in the example_custom != custom[]
and whats worse you can't really tell which type __custom is.

Regarding the referenced issue, I think a better solution would be to cache all arrays instead.
Then type would not change between runs for that test case.

On another note if we want to go full meta, current setup will fail for something like this:

create type integer as (i int)

I would never want a _ for array names. This is internal Postgres thing as demonstrated above.
The main issue is using pg_types as source of information. It would be better if information_schema was used instead. It doesn't suffer from _ prefix issue.

Also, I'm not sure why you think users expect _ as arrays. Postgres dumps always use [], never _ (since that is not correct).